### PR TITLE
Import stats for Live migration

### PR DESCRIPTION
**Pull request by priyanshi-yb**
in continuation of https://github.com/yugabyte/yb-voyager/pull/993

### CODE REVIEW

File: yb-voyager/cmd/live_migration.go [hunk 1]
```diff
@@ -205,23 +210,12 @@ func processEvents(chanNo int, evChan chan *tgtdb.Event, lastAppliedVsn int64, d
...
-			case "c":
-				numInserts++
-			case "u":
-				numUpdates++
-			case "d":
-				numDeletes++
-			}
-		}
-		err := tdb.ExecuteBatch(migrationUUID, tgtdb.EventBatch{Events: batch, ChanNo: chanNo, NumInserts: numInserts, NumUpdates: numUpdates, NumDeletes: numDeletes})
+		eventBatch := &tgtdb.EventBatch{Events: batch, ChanNo: chanNo, EventCounts : &tgtdb.EventCounter{}, EventCountsByTable: make(map[string]*tgtdb.EventCounter)}
```
Comment (amit-yb) [inline yb-voyager/cmd/live_migration.go:213]: Define a new function `eventBatch := tgtdb.NewEventBatch(batch, chanNo)`. The function internally iterates through the `batch` and correctly initialises `EventCounts` and `EventCountsByTable`. With that, you will always have correct batch stats already computed at the time of creation.

File: yb-voyager/src/tgtdb/event.go [hunk 2]
```diff
@@ -253,47 +253,89 @@ func (event *Event) getTableName(targetSchema string) string {
 type EventBatch struct {
 	Events     []*Event
 	ChanNo     int
-	NumInserts int64
-	NumUpdates int64
-	NumDeletes int64
+	EventCounts *EventCounter
+	EventCountsByTable map[string]*EventCounter
 }
```
Comment (amit-yb) [inline yb-voyager/src/tgtdb/event.go:259]: As suggested above implement `NewEventBatch()` here.

File: yb-voyager/src/tgtdb/event.go [hunk 3]
```diff
@@ -253,47 +253,89 @@ func (event *Event) getTableName(targetSchema string) string {
...
+		eb.EventCountsByTable[tableName].NumDeletes, 
+		migrationUUID, tableName, eb.ChanNo)
 }
 
 func (eb *EventBatch) GetTableNames() []string {
-	tableNames := make(map[string]bool)
+	return lo.Keys(eb.EventCountsByTable)
+}
+
+func (eb *EventBatch) GetEventCountsByTable(targetSchema string) map[string]*EventCounter {
```
Comment (amit-yb) [inline yb-voyager/src/tgtdb/event.go:306]: You will need to make this method private and call only from the `NewEventBatch()` to correctly initialise the counters.

File: yb-voyager/src/tgtdb/oracle.go [hunk 4]
```diff
@@ -299,43 +301,77 @@ func (tdb *TargetOracleDB) getLiveMigrationMetaInfoByTable(conn *sql.Conn, migra
...
+	for _, tableName := range tableNames {
+		for c := 0; c < numChans; c++ {
+			tableName = tdb.qualifyTableName(tableName)
+			rowCount, err := tdb.getLiveMigrationMetaInfoByTable(conn, migrationUUID, tableName, c)
+			if err != nil {
+				return  fmt.Errorf("failed to get table wise meta info: %w", err)
+			}
+			if rowCount > 0 {
+				log.Info("table wise meta info already created. Skipping init.")
+			} else {
```
Comment (amit-yb) [inline yb-voyager/src/tgtdb/oracle.go:350]: Shouldn't this be outside of the `numChans` loop?

File: yb-voyager/src/tgtdb/oracle.go [hunk 5]
```diff
@@ -625,8 +653,9 @@ func (tdb *TargetOracleDB) IfRequiredQuoteColumnNames(tableName string, columns
 }
 
 // execute all events sequentially one by one in a single transaction
-func (tdb *TargetOracleDB) ExecuteBatch(migrationUUID uuid.UUID, batch EventBatch) error {
+func (tdb *TargetOracleDB) ExecuteBatch(migrationUUID uuid.UUID, batch *EventBatch) error {
 	// TODO: figure out how to avoid round trips to Oracle DB
+	batch.EventCountsByTable = batch.GetEventCountsByTable(tdb.tconf.Schema)
```
Comment (amit-yb) [inline yb-voyager/src/tgtdb/oracle.go:658]: Generally, you should update fields of a class only from its methods.
If you implement the `NewEventBatch()` as suggested above, you won't need this statement.

File: yb-voyager/src/tgtdb/yugabytedb.go [hunk 6]
```diff
@@ -471,27 +477,62 @@ func (yb *TargetYugabyteDB) initChannelMetaInfo(conn *pgx.Conn, migrationUUID uu
...
+	
+	for _, tableName := range tableNames {
+		for c := 0; c < numChans; c++ {
+			tableName := yb.qualifyTableName(tableName)
+			rowCount, err := yb.getLiveMigrationMetaInfoByTable(conn, migrationUUID, tableName, c)
+			if err != nil {
+				return fmt.Errorf("error getting channels meta info for %s: %w", EVENT_CHANNELS_METADATA_TABLE_NAME, err)
+			}
+			if rowCount > 0 {
+				log.Info(fmt.Sprintf("event stats for %s and channel %d already created. Skipping init.", tableName, c))
```
Comment (amit-yb) [inline yb-voyager/src/tgtdb/yugabytedb.go:510]: Same comment as provided for the oracle implementation.

**Review by amit-yb — COMMENTED** [review]
Few more comments.

File: yb-voyager/src/tgtdb/oracle.go (hunk 4 above)
Comment (priyanshi-yb) [reply yb-voyager/src/tgtdb/oracle.go:350]: which one, `rowCount` check?

File: yb-voyager/src/tgtdb/event.go [hunk 7]
```diff
@@ -257,6 +257,12 @@ type EventBatch struct {
 	EventCountsByTable map[string]*EventCounter
 }
 
+func NewEventBatch(events []*Event, chanNo int, targetSchema string) *EventBatch {
+	batch := &EventBatch{Events: events, ChanNo: chanNo, EventCounts : &EventCounter{}, EventCountsByTable: make(map[string]*EventCounter)}
+	batch.EventCountsByTable = batch.getEventCountsByTable(targetSchema)
```
Comment (amit-yb) [inline yb-voyager/src/tgtdb/event.go:262]: In a single loop you should update both `EventCounts` and `EventCountsByTable`.

**Review by amit-yb — APPROVED** [review]
LGTM with a minor comment.

File: yb-voyager/src/tgtdb/event.go (hunk 7 above)
Comment (priyanshi-yb) [reply yb-voyager/src/tgtdb/event.go:262]: Yes, that is happening in `getEventCountsByTable()` func - 
```
func (eb *EventBatch) getEventCountsByTable(targetSchema string) map[string]*EventCounter {
	counts := make(map[string]*EventCounter)
	for _, event := range eb.Events {
		tableName := event.getTableName(targetSchema)
		if _, ok := counts[tableName]; !ok {
			counts[tableName] = &EventCounter{}
		}
		counts[tableName].CountEvent(event)
		eb.EventCounts.CountEvent(event)
	}
	return counts
}
```

File: yb-voyager/src/tgtdb/event.go [hunk 8]
```diff
@@ -309,17 +314,15 @@ func (eb *EventBatch) GetTableNames() []string {
 	return lo.Keys(eb.EventCountsByTable)
 }
 
-func (eb *EventBatch) getEventCountsByTable(targetSchema string) map[string]*EventCounter {
-	counts := make(map[string]*EventCounter)
+func (eb *EventBatch) getEventCountsByTable(targetSchema string) {
```
Comment (amit-yb) [inline yb-voyager/src/tgtdb/event.go:317]: This is no more a "get" method: how about renaming it as `updateCounters()`

File: yb-voyager/src/reporter/stats/streamImportReporter.go [hunk 9]
```diff
@@ -121,6 +121,6 @@ func (s *StreamImportStatsReporter) BatchImported(numInserts, numUpdates, numDel
 }
 
 func (s *StreamImportStatsReporter) getIngestionRateForLastNMinutes(n int64) int64 {
-	window := 2*n + 1 //2*n as sliding window every 30 secs
+	window := 6*n + 1 //6*n as sliding window every 10 secs
 	return lo.Sum(s.eventsPer30Secs[1:window]) / n
```
Comment (amit-yb) [inline yb-voyager/src/reporter/stats/streamImportReporter.go:125]: Now the `eventsPer30Secs` name is misleading ;-) . How about naming it as just `slidingWindow`?

File: yb-voyager/src/reporter/stats/streamImportReporter.go [hunk 10]
```diff
@@ -121,6 +121,6 @@ func (s *StreamImportStatsReporter) BatchImported(numInserts, numUpdates, numDel
 }
 
 func (s *StreamImportStatsReporter) getIngestionRateForLastNMinutes(n int64) int64 {
-	window := 2*n + 1 //2*n as sliding window every 30 secs
+	window := 6*n + 1 //6*n as sliding window every 10 secs
```
Comment (amit-yb) [inline yb-voyager/src/reporter/stats/streamImportReporter.go:124]: `windowSize`

**Review by amit-yb — APPROVED** [review]
LGTM with minor nits.

File: yb-voyager/src/reporter/stats/streamImportReporter.go (hunk 9 above)
Comment (priyanshi-yb) [reply yb-voyager/src/reporter/stats/streamImportReporter.go:125]: I also just realized that. will rename.

**Pull request merged**